modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -316,6 +325,53 @@ public boolean authenticate(String userName, PublicKey key,
             }
             return successful;
         }
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : Splitter.on(" ").trimResults().omitEmptyStrings().split(line)) {
+                if (linePart.startsWith("from=\"") && linePart.endsWith("\"")) {
+                    String from = linePart.substring(6, linePart.length()-1);
+                    Set<Outcome> outcomes = EnumSet.noneOf(Outcome.class);
+                    for (String pattern : Splitter.on(",").trimResults().omitEmptyStrings().split(from)) {
+                        outcomes.add(patternMatchesHost(pattern, session));
+                    }
+                    return outcomes.contains(Outcome.ALLOW) && !outcomes.contains(Outcome.DENY);
+                }
+            }
+            return true;
+        }
+
+        private Outcome patternMatchesHost(String pattern, ServerSession session) {
+            boolean patternIsNegated = pattern.startsWith("!");
+            Outcome matchOutcome = patternIsNegated ? Outcome.DENY : Outcome.ALLOW;
+            if (patternIsNegated) {
+                pattern = pattern.substring(1);
+            }
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
+                try {
+                    if (addressInCidrRange(pattern, remoteAddress)) {
+                        return matchOutcome;
+                    }
+                } catch (IllegalArgumentException e) {
+                    String remoteName = ((InetSocketAddress) remote).getHostName();
+                    Glob glob = new Glob(pattern);
+                    if (glob.matches(remoteName) || glob.matches(remoteAddress)) {
+                        return matchOutcome;
+                    }
+                }
+            }
+            return Outcome.DEFER;
+        }
+
+        private boolean addressInCidrRange(String cidrAddress, String remoteAddress) throws IllegalArgumentException {
+                try {
+                    Subnet subnet = Subnet.create(cidrAddress);
+                    return subnet.containsHost(remoteAddress);
```

Comment:
Actually, I think you want the `Subnet#contains` method.  This takes an InetAddress (rather than a String).

This has a few benefits: using this method we avoid converting the address into a String form, only to parse it back again.
The method also never throws UnknownHostException :-)

The two lines could be combined into just:

    return Subnet.create(cidrAddress).contains(remoteAddress);

Also, given this method is now just a single line and used just once, you can probably refactor the code to include this line in patternMatchesHost directly:

    if (Subnet.create(cidrAddress).contains(remote.getAddress())) {
        return matchOutcome;
    }